numba/np/ufunc/dufunc.py

```diff
@@ -231,6 +242,264 @@ def _compile_for_argtys(self, argtys, return_type=None):
+                elif init_none:
+                    r = identity
+                else:
+                    r = initial
+
+                sz = array.shape[0]
+                for i in range(start, sz):
+                    r = ufunc(r, array[i])
+                return r
+
+            def impl_nd_axis_int(ufunc,
+                                 array,
+                                 axis=0,
+                                 dtype=None,
+                                 initial=None):
+                if axis is None:
+                    raise ValueError("'axis' must be specified")
+
+                if axis < 0 or axis >= array.ndim:
+                    raise ValueError("Invalid axis")
+
+                # create result array
+                shape = tuple_slice(array.shape, axis)
+
+                if initial is None and identity is None:
+                    r = np.empty(shape, dtype=nb_dtype)
+                    for idx, _ in np.ndenumerate(r):
+                        # shape[0:axis] + 0 + shape[axis:]
+                        result_idx = tuple_slice_append(idx, axis, 0)
+                        r[idx] = array[result_idx]
+                elif initial is None and identity is not None:
+                    # Checking if identity is not none is redundant but required
+                    # compile this block
+                    r = np.full(shape, fill_value=identity, dtype=nb_dtype)
+                else:
+                    r = np.full(shape, fill_value=initial, dtype=nb_dtype)
+
+                # One approach to implement reduce is to remove the axis index
+                # from the indexing tuple returned by "np.ndenumerate". For
+                # instance, if idx = (X, Y, Z) and axis=1, the result index
+                # is (X, Y).
+                # Another way is to compute the result index using strides,
+                # which is faster than manipulating tuples.
+                view = r.ravel()
+                if initial is None and identity is None:
+                    for idx, val in np.ndenumerate(array):
+                        if idx[axis] == 0:
+                            continue
+                        else:
+                            flat_pos = compute_flat_idx(r.strides, r.itemsize,
+                                                        idx, axis)
+                            lhs, rhs = view[flat_pos], val
+                            view[flat_pos] = ufunc(lhs, rhs)
+                else:
+                    for idx, val in np.ndenumerate(array):
+                        if initial is None and identity is None and \
+                                idx[axis] == 0:
+                            continue
+                        flat_pos = compute_flat_idx(r.strides, r.itemsize,
+                                                    idx, axis)
+                        lhs, rhs = view[flat_pos], val
+                        view[flat_pos] = ufunc(lhs, rhs)
+                return r
```

Comment:
The removal of use of `ndenumerate` here as a possible optimisation, should be opened as a separate issue.